File: dev/bloody_rss_man/wp_copier/copier.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="content-type" content="text/html; charset=iso-8859-1" />
  <title>Wordpress copier</title>
</head>
<body>
	<h3>Welcome to copier</h3>

<?php
	
	$rm = getenv('REQUEST_METHOD');
	
	$input_dest = 'destname';
	$input_database = 'dbname';
	
	$base_dir = '~/public_html';
	$wp_base = $base_dir . '/1wp';
	$themes_base = $base_dir . '/1themes';
	
	$wp_config = $wp_base . '/wp-config.php';
	
	if ($rm=='POST'){
		$dest = $_REQUEST[$input_dest];
		$db = $_REQUEST[$input_database];

		$ok = false;
		
		if($db){
			
			if(file_exists($wp_config)){
				
				# read configuration
				$config = read_file($wp_config);
				
				# change database name
				$pattern = "/define\('DB_NAME',\s*'.*?'\);/";
				$replacement = "define('DB_NAME', '$db');";
				write_file($wp_config, preg_replace($pattern, $replacement, $config));
				echo "<i>Database name was changed to $db</i><br/>";
				
				# copy wordpress
				
				$wp_source = $wp_base . '/*';
				$wp_dest = "$base_dir/$dest";
				
				# check existence of the destination folder
				if(file_exists($wp_dest) && is_dir($wp_dest)){

					# do copy wordpress
					if(do_exec("cp -r $wp_source $wp_dest")){
						echo "<i>Wordpress was copied to $wp_dest</i><br/>";
						# copied, let's copy a theme
						
						# get all themes
						$theme_dir = opendir($themes_base);
						$has_themes = false;
						while($file = readdir($theme_dir)){
							if ($file!='.' && $file!='..' && is_dir("$themes_base/$file")){
								$list[] = $file;
								$has_themes = true;
							}
						}
						closedir($theme_dir);
						
						if($has_themes){
							
							shuffle($list);
							$theme = $list[0];
							
							# move the theme to the target folder
							echo "<i>Using theme $theme</i><br/>";
							
							if(do_exec("cp $themes_base/$theme $wp_dest/wp-content/themes")){
								
								
								
								$ok = true;
							}
						} else {
							echo "<b>No themes</b><br/>";
						}
						
					}
				} else {
					echo "<b>No target directory $wp_dest</b><br/>";
				}
			} else {
				echo "<b>No file $wp_config</b><br/>";
			}
		} else {
			echo "<b>Please specify the database name</b><br/>";
		}
		
		if ($ok)
			echo "Your request was successfully processed.<br/>";
		
		echo "<br/><br/><a href=".">Back to copier</a>";
		
	} else {
		echo "<form name='data' method='post'>";

			echo "Enter name ot the destination directory under the '/public_html':<br/>";
			echo "<input type='text' name='$input_dest' /><br/><br/>";
		
			echo "Enter database name:<br/>";
			echo "<input type='text' name='$input_database' /><br/><br/>";
			
			echo "<input type='submit' value=' Ok ' />";
		
		echo "</form>";
	}

function do_exec($cmd) {
	$code;
	$output;
	exec($cmd, &$output, &$code);
	if($code!=0){
		echo "<b>Cannot execute $cmd:</b><br/><pre>";
		echo implode("<br/>", $output);
		echo "</pre><br/>";
		return false;
	} else
		return true;
}

function write_file($filename, $newdata) {
	$f = fopen($filename, "w");
	fwrite($f, $newdata);
	fclose($f);  
}

function read_file($filename) {
	$f = fopen($filename, "r");
	$data=fread($f, filesize($filename));
	fclose($f);  
	return $data;
}

?>

</body>
</html>

File: dev/bloody_rss_man/wp_copier/copier.pl
use strict;
use warnings;

use CGI ();
CGI->compile(':all');
use CGI::Carp qw(fatalsToBrowser);

my $xcgi = new CGI;
start($xcgi);

sub start {
	my $cgi = shift;

	my $toname = 'copy_to';
	my $dbname = 'dbname';
	
	my $respname = 'isresponse';
	
	print $cgi->header,
		$cgi->start_html('Copier'),
		$cgi->h2('Welcome to the copier');
	
	
	# do operation
	my @output;
	my $ok;
	
	if($cgi->param($respname)){
		# 1. update db name
		$ok = update_db_name($cgi, $dbname, \@output);
		
		# 2. copy wordpress
		$ok = $ok && copy_wordpress($cgi, $toname, \@output);
		
		# 3. get and move a theme
		$ok = $ok && copy_theme($cgi, $toname, \@output);
		
		if($ok){
			$cgi->param($dbname, '');
			$cgi->param($toname, '');
		}
	}
		
	if(@output>0){
		print $cgi->hr;
		print join "<br/>", @output;
		print $cgi->hr;
		print "<br/>";
	}

	# print form
  print $cgi->start_form('post');
	
		print "Copy files from /public_html/wp1 to:<br/>";
		print $cgi->textfield($toname, ''), "<br/><br/>";
		
		print "Change database name to<br/>";
		print $cgi->textfield($dbname, ''), "<br/><br/>";

		print $cgi->submit('Sumbit');
		
		print $cgi->hidden($respname, 1);
		
	print $cgi->endform;

	
	
	
	print $cgi->end_html;
}


sub update_db_name {
	my ($cgi, $paramname, $logref) = @_;
	my $name = $cgi->param($paramname);
	if($name){
		
		
		push @$logref, "<i>Database name was changed to $name</i>";
		return 1;
	} else {
		push @$logref, "<i><b>Database name cannot be empty</b></i>";
		return 0;
	}
	
}

sub copy_wordpress {
	my ($cgi, $paramname, $logref) = @_;
	my $path = $cgi->param($paramname);
	$path =~ s!/+$!!;
	if($path){
		
		
		
		my $wp_source = "~/public_html/1wp";
		my $wp_dest = "~/public_html/$path";
		
		push @$logref, "<i>Wordpress was copied to $path</i>";
		return 1;
	} else {
		push @$logref, "<i><b>Copy path cannot be empty</b></i>";
		return 0;
	}
}

sub copy_theme {
	my ($cgi, $paramname, $logref) = @_;
	
	my $path_to = $cgi->param($paramname);
	
	# get awailable themes
	my $themes_path = '/public_html/1themes';
	opendir (DIR, $themes_path) or die "Cannot open directory $themes_path: $!";
	my @themes = grep { /[^.]/ && -d "$themes_path/$_" } readdir DIR;
	closedir DIR;
	
	if(@themes==0){
		push @$logref, "<i><b>No themes!</b></i>";
		return 0;
	}
	
	srand;
	my $ti = int(rand(scalar @themes));
	
	push @$logref, "Found themes: @themes";
	push @$logref, "Copy theme $themes[$ti] to $path_to";
	
}




1;